Add minimap for wxSTC

vadz opened the pull request.

vadz committed:
Avoid using Tab and Shift+Tab as accelerators

This doesn't work in wxGTK and results in a debug warning on the sample
startup.

vadz committed:
Make wxStyledTextCtrl::TextHeight() const

This function isn't declared as "get" in the Scintilla interface file
but it's still one, semantically, so make it const in C++.

vadz committed:
Experiment with document map in wxSTC sample

Add a menu item showing a dialog with the document and its map.

vadz committed:
Make the overlay more opaque while dragging under non-MSW too

Unfortunately opacity needs to be handled differently for wxMSW and the
others.

vadz committed:
Prevent Scintilla from handling double clicks in the map

This results in unwanted text selection and messes up mouse capture, so
don't let it process this event.

vadz committed:
Wrap lines in the main document used with the map for testing

vadz committed:
Don't scroll when starting to drag

Add forgotten return to prevent falling through to the logic for
clicking outside of the thumb when clicking on it.

vadz committed:
Add convenient GetThumbTopPos() function

Just extract it from DrawVisibleZone().

No real changes, this is just a refactoring.

vadz committed:
Simplify and improve map thumb positioning logic

Cache often used values for efficiency.

Avoid recursive calls to our Sync{Edit,Map}Position() by using a flag to
block the unwanted notifications.

Fix thumb positioning while dragging.

vadz committed:
Handle double clicks in the map as simple clicks

This is much more useful than ignoring them as some clicks in a series
of clicks could be unintentionally lost/ignored before.

vadz committed:
Don't allow Ctrl+mouse wheel zooming in the map

Still allow scrolling using just the mouse wheel.

vadz committed:
Prevent removing the document or the map in ShowDocumentMap()

Don't allow un-splitting wxSplitterWindow.

vadz committed:
Store backing store bitmap in wxStyledTextCtrl

Avoid recreating this bitmap in each call to OnPaint().

vadz committed:
Use wxBitmap object and not pointer in SurfaceImpl

There is no need to dynamically allocate this object when we can just
use null bitmap instead of null pointer.

Using the object is simpler, safer and could be marginally faster.

vadz committed:
Store the last 8 used brushes in wxSTC implementation

This avoid constantly recreating the background brush which is used
multiple times on every repaint.

vadz committed:
Reuse the same string buffer for STC-to-wx string conversions

This avoids a lot of memory allocation while redrawing the window and
results in noticeable performance improvement.

vadz committed:
Avoid casting arbitrary wxDC to wxMemoryDC in wxSTC

Even though all the device contexts used here are currently of either
wxMemoryDC or wxBufferedPaintDC (deriving from wxMemoryDC) type, casting
them to wxMemoryDC is still dangerous and prevents from the code being
ever changed to use other wxDC types.

No real changes, just make the code safer.

vadz committed:
Disable bounding box updates for wxDCs used in wxSTC

This is another micro-optimization for the drawing code.

vadz committed:
Merge branch 'master' into stc-document-map

Pull in the performance-related changes in order to use them in this
branch.

vadz committed:
Make VisibleFromDocLine/DocLineFromVisible() const too

These functions names don't start with "Get" but they still don't modify
the control and so should be const.

vadz committed:
Use variable instead of hardcoded number for folding margin

Make the comparison more understandable and avoid hardcoding "2".

vadz committed:
Do not set the lexer for the map wxSTC instance

Apparently using the same lexer for 2 controls sharing the same document
doesn't work, this results in problems with syntax highlighting and
folding not working at all.

Simply not doing this fixes these problems while still showing the
correct syntax highlighting in the map, as it shares the same document.

vadz committed:
Use number of displayed, and not document, lines in the map code

We really care about the number of physical, i.e. "display", lines
appearing in the window and not the number of logical, or "document",
lines.

They are the same as long as no lines are either wrapped or folded but
become different as soon as either of them is used.

vadz committed:
Add wxStyledTextCtrl::GetDisplayLineCount()

This is a convenient, even if trivial, function.

vadz committed:
Merge branch 'master' into stc-document-map

vadz committed:
Add separate m_isDragging instead of reusing m_dragOffset

The value of m_dragOffset will possibly become negative after the
upcoming changes, so don't rely on the special value -1 meaning that we
are not dragging any more.

No real changes.

vadz committed:
Initialize m_mapLineHeight only once

It doesn't change even if the window is resized.

No real changes.

vadz committed:
Collect information about visible lines in LinesInfo struct

This makes it simpler to check if any of the relevant variables have
changed and update them all at once.

vadz committed:
Rename GetMapLineAtPoint() to GetDocLineAtMapPoint()

Make it clear that this function returns the logical/document line and
not the physical/display line.

No real changes.

vadz committed:
Sync map and document properly when lines are wrapped

Don't assume that display and document lines are the same any longer.

This complicates things a lot, notably while dragging because we can't
simply find the correct first line for the editor any longer but need to
search for it iteratively because the size of the thumb now depends on
its position, as different regions of the document may wrap differently.

Also synchronize the map with the document when it is repainted and not
resized because line wrapping is done in the background after the resize
and we don't have the correct mapping between the display and document
lines yet in the wxEVT_SIZE handler -- but we eventually get it for the
last wxEVT_PAINT provoked by it.

vadz committed:
Merge remote-tracking branch 'origin/master' into stc-document-map

vadz committed:
Rename all "editor" variables to start with "edit" instead

Use consistent naming instead of using "editorFoo" for some things and
"editBar" for the other ones.

No real changes.

vadz committed:
Fix multiple bugs in determining position when dragging

Improve the initial estimate: old method was more precise when display
and document lines matched up (i.e. without wrapping) but could result
in completely wrong value when they were very different.

Add proper range checking to avoid ending up with an invalid editor
first visible line.

Ensure that we don't enter an infinite loop when there are two first
visible lines values resulting in the thumb position close to but
either just above or below the correct one.

vadz committed:
Merge branch 'master' into stc-document-map

vadz committed:
Replace {map,edit}Display with {map,edit}Max in LinesInfo

No real changes, just store the maximum possible values for the first
visible line in both controls instead of storing the number of display
lines as we only really need the former.

vadz committed:
Improve skipping scroll events caused by the program itself

Don't use a single m_updateBlocked flag which was reset too early,
before the re-wrapping happened, and so didn't really avoid processing
the scroll events due to the program itself.

Remember the last first line value we set instead and ignore events
scrolling to this line instead.

vadz committed:
Really fix map sync when lines can be wrapped

The assumption that first visible lines of the map and the editor are
proportional doesn't hold in this case, only the thumb position and the
first visible line in the editor must be proportional to each other in
order to accurately represent the position in the document.

vadz committed:
Add MapVisibleFromEditVisible() helper function

No real changes, just simplify the code a bit by adding a trivial
function performing a commonly needed mapping.

vadz committed:
Fix scrolling the editor while dragging the thumb

Find the candidate middle of the thumb directly from the first visible
line in the editor, we don't need to compute the first visible map line
(especially as we did it wrongly) for this.

Get rid of m_dragLastY which is not needed to get a reasonably accurate
initial estimation.

vadz committed:
Fix SyncEditPosition() for the case of wrapped lines too

We need to iteratively find the first editor line such that the
corresponding first map line is the desired one.

vadz committed:
Extract iteration algorithm in its own function

Avoid duplicating the same code in different functions that need to
iteratively find the correct first visible line in the editor.

vadz committed:
Increase the line number margin width

It wasn't big enough to show 4 digit numbers.

vadz committed:
Add wxStyledTextCtrl::SetMirrorFoldingCtrl()

This can be used to mirror folding in one control to another one and
will be used by wxStyledTextCtrlMiniMap to keep its folds synchronized
with those in the main control automatically.

vadz committed:
Synchronize folding in the map and the main editor

Also don't hide fold margin any longer to allow testing folding.

vadz committed:
Add Edit::ToggleLineNumbers() to the sample

Make it a bit simpler to show line numbers when showing the minimap.

vadz committed:
Move wxStyledTextCtrlMiniMap from stc sample to the library

After experimenting with this code in the sample, move it to its
permanent place in the library.

This commit is best viewed using Git --color-moved option and ignoring
whitespace changes.

vadz committed:
Add documentation for wxStyledTextCtrlMiniMap

There is not much to document, as this class needs to be just created
and left to its own devices.

eranif commented:
Nice! Looking forward to use this.

vadz committed:
Add possibility to customize wxStyledTextCtrlMiniMap thumb colours

Default grey may not be always appropriate.

vadz committed:
Don't scroll wxStyledTextCtrl used by wxStyledTextCtrlMiniMap

Scrolling the control using wxWindow::ScrollWindow() corrupted the thumb
drawn on top of it, as parts of it were scrolled away. There doesn't
seem to be any way to prevent this from happening, so make it possible
to avoid using ScrollWindow() completely by adding new SetCustomDrawn()
function and using it in wxStyledTextCtrlMiniMap.

This fixes visual thumb corruption under wxMSW when scrolling the map by
small increments, e.g. using mouse wheel or the scrollbar arrows.

vadz commented:
> BTW, there is a small drawing glitch on Windows 11: If you scroll the "Document Map" using the scrollbar, the gray area has some drawing artifacts.

Just to confirm: this is with the latest state of this branch, including c39b5d9494937aafb7bfefe877394ccb186b91c6? Because it was supposed to fix exactly this (and did, in my testing...).

eranif commented:
I am missing this commit (I applied this patch locally by applying: https://github.com/wxWidgets/wxWidgets/pull/25887.diff and this was done before you added that commit).

vadz commented:
Actually, looking at your patch, you definitely don't have this commit in your tree (because it doesn't apply to the header with this change), please get the latest branch and the refresh problem should go away.

vadz commented:
> It works.
> 
> Below is a diff of the changes I did.

Thanks for checking, but the diff can't be used as is, unfortunately: it modifies generated files and so the changes will be lost the next time `src/stc/gen_iface.py` is run.

The straightforward way of doing it properly would be to modify this script to add the custom function definitions to it just as I did for folding, but looking at the number of functions, it doesn't seem very appealing, so I think that it would be better to let the script know about the mirrored functions in some way, so that it could add mirroring code automatically. I don't have time to do it right now, but I'll try to do it later if nobody else beats me to it.

Also, it's not great that all the markers defined before establishing the mirror relationship have to be manually applied to the map. Could we use `MarkerNext()` to get all the existing markers and apply them automatically? If not, we at least need to document this carefully.

eranif commented:
> Also, it's not great that all the markers defined before establishing the mirror relationship have to be manually applied to the map. Could we use MarkerNext() to get all the existing markers and apply them automatically? If not, we at least need to document this carefully

It is doable. We should also need to "copy" all the marker definitions as well.

vadz committed:
Refresh the map when syncing its position even when not scrolling

If the entire map fits into the map window, the map first visible line
never changes as it always remains as 0, however we still need to
refresh the map to show the new thumb position properly.

vadz committed:
Don't sync edit after map changes originating from the editor

If the change in the map scrolling position is due to a change in the
editor contents, we don't need to synchronize the editor with the map
and, in fact, it is wrong to do it as deleting text at the bottom of the
editor could result in the editor being scrolled to the top if the map
was scrolled to the top due to it fitting the window now when it didn't
fit it before.

vadz committed:
Rename wxSTC::SetMirrorFoldingCtrl() to SetMirrorCtrl()

This function will be used to synchronize other changes than just the
folds in the upcoming commit, so stop using "Folding" in its name.

vadz committed:
Make generating of mirroring methods more maintainable

Instead of requiring to provide the full implementation of all methods
that must be forwarded to m_mirrorCtrl, define a list of such methods
and handle all of them in a single place.

This will make adding new methods to mirror much simpler.

vadz committed:
Synchronize markers with the mirror control too

Add a demo marker to the sample to verify that it is shown in the map as
well.

vadz committed:
Merge branch 'master' into stc-document-map

Resolve conflicts in Xcode projects by regenerating them.

vadz committed:
Don't explicitly refresh the map when the editor contents change

This seems to be unnecessary, as the map contents is updated anyhow, and
just results in extra flicker as seen when e.g. maintaining a key
pressed on autorepeat in the editor.

vadz committed:
Don't synchronize map position on resize to avoid unwanted jumps

Map position may change when its size changes, but this shouldn't result
in a change in the editor position, as it did before this change because
calling SyncMapPosition() from wxEVT_SIZE handler resulted in a
subsequent call to SyncEditPosition() when we received a V_SCROLL update
from Scintilla.

Stop doing this from this handler and instead remember the last "set"
(actually, "known") map position to avoid reacting to the corresponding
V_SCROLL change.

Note that just setting m_lastSetMapFirst but continuing to call
SyncMapPosition() from here didn't quite work because Scintilla seems to
not be in a consistent state yet when this event handler is executed
when maximizing the window containing the map and the map position could
pass by some small intermediate value (2 seen in the sample) before
finally jumping to 0, which prevented us from filtering this update out.

Updating map only from wxEVT_PAINT handler doesn't seem to result in any
visual problems and avoids this problem.

vadz committed:
Don't hide editor vertical scrollbar when using minimap

This looks nicer but loses potentially useful functionality of the
editor scrollbar which may be used to scroll the editor even if the text
is short enough to fit entirely in the map, and so when the map doesn't
have any scrollbar at all.

Document the possibility of hiding it scrollbar for people who still
want to do it.

vadz committed:
Add wxStyledTextCtrlMiniMap::SetEdit()

Allow changing the editor associated with the map or even resetting it
completely, so that the map doesn't show anything.

Also make the relationship between the editor and the map safer, by
avoid leaving dangling pointers when either of them is destroyed.

vadz committed:
Fix document pointer memory leak in wxStyledTextCtrlMiniMap

Remove the extraneous call to AddRefDocument(), SetDocPointer() already
does this internally and doing this in wxStyledTextCtrlMiniMap code just
ensured that we leaked the associated document.

vadz committed:
Document limitations of wxStyledTextCtrlMiniMap synchronization

Mention that only folds and markers updated after associating the map
with the control are automatically propagated to it and that the
application needs to do it manually for the pre-existing ones.

eranif commented:
I saw this crash on **macOS** as well, however, I was not able to reproduce it with **cmake** builds (I am not sure how it's related...). 

The crash happened in `MeasuredWidth`, somewhere in `wxArrayInt.Add` or something like that... building the `stctest` sample with `configure` on macOS crashes immediately with this crash. I think I still have an ASAN build laying around so I can provide its output if needed.

vadz commented:
I am completely stumped as I don't see this here, don't see any problems in ASAN build under Linux and no problems when running normal build under Valgrind and, most of all, because of this

> Or change it to stcString = wxString::FromUTF8Unchecked(s, len);

as this really shouldn't change anything at all (the other fix, of returning a copy of the string, undoes the whole point of this function which is to avoid memory allocations).

I thought that perhaps there was some problem due to reentrancy, but applying this:

```diff
diff --git a/src/stc/PlatWX.cpp b/src/stc/PlatWX.cpp
index 2ab1e707cd..8cd5c5bc3e 100644
--- a/src/stc/PlatWX.cpp
+++ b/src/stc/PlatWX.cpp
@@ -214,7 +214,12 @@ private:
     // avoid reallocating memory all the time, e.g. for each line drawn.
     static wxString stcString;
 
+    static int stcStringInsideFlag;
+
     const wxString& STCString(const char* s, int len) {
+        wxRecursionGuard guard(stcStringInsideFlag);
+        wxASSERT( !guard.IsInside() );
+
         stcString.AssignFromUTF8Unchecked(s, len);
 
         return stcString;
@@ -275,6 +280,7 @@ public:
 
 std::array<SurfaceImpl::BrushCacheEntry, 8> SurfaceImpl::brushCache;
 wxString SurfaceImpl::stcString;
+int SurfaceImpl::stcStringInsideFlag = 0;
 
 
 SurfaceImpl::SurfaceImpl() :
```

doesn't result in any asserts, so it's not this either (or maybe I just can't get into a situation in which this happens, which would also explain why I don't see it).

@eranif Please do post ASAN output if you have it, TIA!

eranif commented:
I could not find it, so I re-built with ASAN and the `stctest` sample crashes instantly.

The command line I used for the build:

```bash
CFLAGS="-fsanitize=address -g -fno-omit-frame-pointer" CXXFLAGS="-fsanitize=address -g -fno-omit-frame-pointer" LDFLAGS="-fsanitize=address" ../configure --enable-shared --enable-monolithic --with-osx_cocoa --enable-debug --disable-mediactrl --enable-stl --with-libtiff=no --enable-utf8
```

Running the sample directly (without `open`):

```bash
$ stctest.app/Contents/MacOS/stctest
stctest(55160,0x2040ae0c0) malloc: nano zone abandoned due to inability to reserve vm space.
=================================================================
==55160==ERROR: AddressSanitizer: container-overflow on address 0x61600006ebf8 at pc 0x00010565cdac bp 0x00016fc822b0 sp 0x00016fc822a8
READ of size 8 at 0x61600006ebf8 thread T0
    #0 0x00010565cda8 in wxGCDCImpl::DoGetPartialTextExtents(wxString const&, wxArrayInt&) const dcgraph.cpp:1302
    #1 0x0001056229ac in wxReadOnlyDC::GetPartialTextExtents(wxString const&, wxArrayInt&) const dc.h:874
    #2 0x00010663a0a8 in Scintilla::SurfaceImpl::MeasureWidths(Scintilla::Font&, char const*, int, float*) PlatWX.cpp:715
    #3 0x000106bfc8e8 in Scintilla::PositionCache::MeasureWidths(Scintilla::Surface*, Scintilla::ViewStyle const&, unsigned int, char const*, unsigned int, float*, Scintilla::Document const*) PositionCache.cxx:811
    #4 0x000106b02e80 in Scintilla::EditView::LayoutLine(Scintilla::EditModel const&, long, Scintilla::Surface*, Scintilla::ViewStyle const&, Scintilla::LineLayout*, int) EditView.cxx:490
    #5 0x000106b24a54 in Scintilla::EditView::PaintText(Scintilla::Surface*, Scintilla::EditModel const&, Scintilla::PRectangle, Scintilla::PRectangle, Scintilla::ViewStyle const&) EditView.cxx:2294
    #6 0x000106b50d58 in Scintilla::Editor::Paint(Scintilla::Surface*, Scintilla::PRectangle) Editor.cxx:1797
    #7 0x0001066659e4 in ScintillaWX::DoPaint(wxDC*, wxRect) ScintillaWX.cpp:854
    #8 0x0001065fe77c in wxStyledTextCtrl::OnPaint(wxPaintEvent&) stc.cpp:5551
    #9 0x000104cc8710 in wxAppConsoleBase::HandleEvent(wxEvtHandler*, void (wxEvtHandler::*)(wxEvent&), wxEvent&) const appbase.cpp:663
    #10 0x000104cc88a8 in wxAppConsoleBase::CallEventHandler(wxEvtHandler*, wxEventFunctor&, wxEvent&) const appbase.cpp:675
    #11 0x00010514b3ec in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) event.cpp:1468
    #12 0x00010514ad34 in wxEventHashTable::HandleEvent(wxEvent&, wxEvtHandler*) event.cpp:1074
    #13 0x000105150bfc in wxEvtHandler::TryHereOnly(wxEvent&) event.cpp:1655
    #14 0x0001051505e8 in wxEvtHandler::TryBeforeAndHere(wxEvent&) event.h:4078
    #15 0x000105150628 in wxEvtHandler::ProcessEventLocally(wxEvent&) event.cpp:1588
    #16 0x0001051502cc in wxEvtHandler::ProcessEvent(wxEvent&) event.cpp:1561
    #17 0x00010516a258 in wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_0::operator()() const event.cpp:1675
    #18 0x0001051518a4 in bool wxSafeCall<bool, wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_0, wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_1>(wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_0 const&, wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_1 const&) safecall.h:40
    #19 0x00010514f8a4 in wxEvtHandler::SafelyProcessEvent(wxEvent&) event.cpp:1673
    #20 0x000105950d38 in wxWindowBase::HandleWindowEvent(wxEvent&) const wincmn.cpp:1517
    #21 0x00010528981c in wxWindow::MacDoRedraw(long) window_osx.cpp:2006
    #22 0x000105522e8c in wxWidgetCocoaImpl::drawRect(void*, NSView*, void*)+0x3f0 (libwx_osx_cocoau-3.3.2.0.0.dylib:arm64+0x88ae8c)
    #23 0x00010551ec00 in wxOSX_drawRect(NSView*, objc_selector*, CGRect)+0x1c0 (libwx_osx_cocoau-3.3.2.0.0.dylib:arm64+0x886c00)
    #24 0x00019a23cca4 in _NSViewDrawRect+0x9c (AppKit:arm64e+0x14dca4)
    #25 0x00019ac53114 in -[NSView _recursive:displayRectIgnoringOpacity:inContext:stopAtLayerBackedViews:]+0x438 (AppKit:arm64e+0xb64114)
    #26 0x00019a23c698 in -[NSView(NSLayerKitGlue) _drawViewBackingLayer:inContext:drawingHandler:]+0x214 (AppKit:arm64e+0x14d698)
    #27 0x00019a8f9b8c in -[NSViewBackingLayer drawInContext:]+0x34 (AppKit:arm64e+0x80ab8c)
    #28 0x00019a56fa78  (AppKit:arm64e+0x480a78)
    #29 0x00019a557f9c  (AppKit:arm64e+0x468f9c)
    #30 0x00019a553110  (AppKit:arm64e+0x464110)
    #31 0x00019a557c6c  (AppKit:arm64e+0x468c6c)
    #32 0x00019a555670  (AppKit:arm64e+0x466670)
    #33 0x00019a5e0534  (AppKit:arm64e+0x4f1534)
    #34 0x00019a5e0b9c  (AppKit:arm64e+0x4f1b9c)
    #35 0x00019a8f97ac in -[NSViewBackingLayer display]+0x4fc (AppKit:arm64e+0x80a7ac)
    #36 0x00019f2840b0 in CA::Layer::display_if_needed(CA::Transaction*)+0x2e8 (QuartzCore:arm64e+0x210b0)
    #37 0x00019f418460 in CA::Context::commit_transaction(CA::Transaction*, double, double*)+0x1e8 (QuartzCore:arm64e+0x1b5460)
    #38 0x00019f2653a4 in CA::Transaction::commit()+0x280 (QuartzCore:arm64e+0x23a4)
    #39 0x00019a24bfb0 in __62+[CATransaction(NSCATransaction) NS_setFlushesWithDisplayLink]_block_invoke+0x10c (AppKit:arm64e+0x15cfb0)
    #40 0x00019ac95974 in ___NSRunLoopObserverCreateWithHandler_block_invoke+0x3c (AppKit:arm64e+0xba6974)
    #41 0x000196205ed4 in __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+0x20 (CoreFoundation:arm64e+0x7bed4)
    #42 0x000196205dbc in __CFRunLoopDoObservers+0x214 (CoreFoundation:arm64e+0x7bdbc)
    #43 0x00019620542c in __CFRunLoopRun+0x30c (CoreFoundation:arm64e+0x7b42c)
    #44 0x000196204a94 in CFRunLoopRunSpecific+0x238 (CoreFoundation:arm64e+0x7aa94)
    #45 0x0001a1ca7278 in RunCurrentEventLoopInMode+0x140 (HIToolbox:arm64e+0xc3278)
    #46 0x0001a1caa318 in ReceiveNextEventCommon+0xd4 (HIToolbox:arm64e+0xc6318)
    #47 0x0001a1e35480 in _BlockUntilNextEventMatchingListInModeWithFilter+0x48 (HIToolbox:arm64e+0x251480)
    #48 0x00019a129a30 in _DPSNextEvent+0x2a8 (AppKit:arm64e+0x3aa30)
    #49 0x00019aac893c in -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:]+0x2ac (AppKit:arm64e+0x9d993c)
    #50 0x00019a11cbe0 in -[NSApplication run]+0x1dc (AppKit:arm64e+0x2dbe0)
    #51 0x0001054eee10 in wxGUIEventLoop::OSXDoRun()+0x9c (libwx_osx_cocoau-3.3.2.0.0.dylib:arm64+0x856e10)
    #52 0x0001050a248c in wxCFEventLoop::DoRun() evtloop_cf.cpp:326
    #53 0x000104dd442c in wxEventLoopBase::Run() evtloopcmn.cpp:88
    #54 0x000104cc59b0 in wxAppConsoleBase::MainLoop() appbase.cpp:380
    #55 0x000104cc4bec in wxAppConsoleBase::OnRun() appbase.cpp:291
    #56 0x000105593a00 in wxAppBase::OnRun() appcmn.cpp:347
    #57 0x000105381940 in wxApp::OnRun() app.cpp:387
    #58 0x000104e9f7cc in wxEntry(int&, wchar_t**)::$_0::operator()() const init.cpp:576
    #59 0x000104e9b960 in int wxSafeCall<int, wxEntry(int&, wchar_t**)::$_0, wxEntry(int&, wchar_t**)::$_1>(wxEntry(int&, wchar_t**)::$_0 const&, wxEntry(int&, wchar_t**)::$_1 const&) safecall.h:40
    #60 0x000104e9b444 in wxEntry(int&, wchar_t**) init.cpp:557
    #61 0x000104e9baf8 in wxEntry(int&, char**) init.cpp:590
    #62 0x000100174ee4 in main stctest.cpp:191
    #63 0x000195d7ab94 in start+0x17b8 (dyld:arm64e+0xfffffffffff3ab94)

0x61600006ebf8 is located 376 bytes inside of 624-byte region [0x61600006ea80,0x61600006ecf0)
allocated by thread T0 here:
    #0 0x000100beb404 in _Znwm+0x74 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x4b404)
    #1 0x0001053bf7d4 in void* std::__1::__libcpp_operator_new[abi:ne200100]<double>(unsigned long) allocate.h:55
    #2 0x0001053bf7b0 in double* std::__1::__libcpp_allocate[abi:ne200100]<double>(std::__1::__element_count, unsigned long) allocate.h:91
    #3 0x0001053bf77c in std::__1::allocator<double>::allocate[abi:ne200100](unsigned long) allocator.h:105
    #4 0x0001053bf714 in std::__1::__allocation_result<std::__1::allocator_traits<std::__1::allocator<double>>::pointer> std::__1::__allocate_at_least[abi:ne200100]<std::__1::allocator<double>>(std::__1::allocator<double>&, unsigned long) allocate_at_least.h:41
    #5 0x0001053bf584 in std::__1::__split_buffer<double, std::__1::allocator<double>&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<double>&) __split_buffer:325
    #6 0x0001053bf0fc in std::__1::__split_buffer<double, std::__1::allocator<double>&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<double>&) __split_buffer:321
    #7 0x0001053b5c74 in std::__1::vector<double, std::__1::allocator<double>>::reserve(unsigned long) vector.h:1086
    #8 0x0001053b557c in wxMacCoreGraphicsContext::GetPartialTextExtents(wxString const&, wxArrayDouble&) const graphics.cpp:2628
    #9 0x00010565cd1c in wxGCDCImpl::DoGetPartialTextExtents(wxString const&, wxArrayInt&) const dcgraph.cpp:1300
    #10 0x0001056229ac in wxReadOnlyDC::GetPartialTextExtents(wxString const&, wxArrayInt&) const dc.h:874
    #11 0x00010663a0a8 in Scintilla::SurfaceImpl::MeasureWidths(Scintilla::Font&, char const*, int, float*) PlatWX.cpp:715
    #12 0x000106bfc8e8 in Scintilla::PositionCache::MeasureWidths(Scintilla::Surface*, Scintilla::ViewStyle const&, unsigned int, char const*, unsigned int, float*, Scintilla::Document const*) PositionCache.cxx:811
    #13 0x000106b02e80 in Scintilla::EditView::LayoutLine(Scintilla::EditModel const&, long, Scintilla::Surface*, Scintilla::ViewStyle const&, Scintilla::LineLayout*, int) EditView.cxx:490
    #14 0x000106b24a54 in Scintilla::EditView::PaintText(Scintilla::Surface*, Scintilla::EditModel const&, Scintilla::PRectangle, Scintilla::PRectangle, Scintilla::ViewStyle const&) EditView.cxx:2294
    #15 0x000106b50d58 in Scintilla::Editor::Paint(Scintilla::Surface*, Scintilla::PRectangle) Editor.cxx:1797
    #16 0x0001066659e4 in ScintillaWX::DoPaint(wxDC*, wxRect) ScintillaWX.cpp:854
    #17 0x0001065fe77c in wxStyledTextCtrl::OnPaint(wxPaintEvent&) stc.cpp:5551
    #18 0x000104cc8710 in wxAppConsoleBase::HandleEvent(wxEvtHandler*, void (wxEvtHandler::*)(wxEvent&), wxEvent&) const appbase.cpp:663
    #19 0x000104cc88a8 in wxAppConsoleBase::CallEventHandler(wxEvtHandler*, wxEventFunctor&, wxEvent&) const appbase.cpp:675
    #20 0x00010514b3ec in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) event.cpp:1468
    #21 0x00010514ad34 in wxEventHashTable::HandleEvent(wxEvent&, wxEvtHandler*) event.cpp:1074
    #22 0x000105150bfc in wxEvtHandler::TryHereOnly(wxEvent&) event.cpp:1655
    #23 0x0001051505e8 in wxEvtHandler::TryBeforeAndHere(wxEvent&) event.h:4078
    #24 0x000105150628 in wxEvtHandler::ProcessEventLocally(wxEvent&) event.cpp:1588
    #25 0x0001051502cc in wxEvtHandler::ProcessEvent(wxEvent&) event.cpp:1561
    #26 0x00010516a258 in wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_0::operator()() const event.cpp:1675
    #27 0x0001051518a4 in bool wxSafeCall<bool, wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_0, wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_1>(wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_0 const&, wxEvtHandler::SafelyProcessEvent(wxEvent&)::$_1 const&) safecall.h:40
    #28 0x00010514f8a4 in wxEvtHandler::SafelyProcessEvent(wxEvent&) event.cpp:1673
    #29 0x000105950d38 in wxWindowBase::HandleWindowEvent(wxEvent&) const wincmn.cpp:1517

HINT: if you don't care about these errors you may set ASAN_OPTIONS=detect_container_overflow=0.
If you suspect a false positive see also: https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow.
SUMMARY: AddressSanitizer: container-overflow dcgraph.cpp:1302 in wxGCDCImpl::DoGetPartialTextExtents(wxString const&, wxArrayInt&) const
Shadow bytes around the buggy address:
  0x61600006e900: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x61600006e980: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa
  0x61600006ea00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61600006ea80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x61600006eb00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x61600006eb80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00[fc]
  0x61600006ec00: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc
  0x61600006ec80: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fa fa
  0x61600006ed00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61600006ed80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x61600006ee00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==55160==ABORTING
Abort trap: 6
```
